zebra-consensus/src/checkpoint.rs

```diff
@@ -754,16 +795,74 @@ impl Service<Arc<Block>> for CheckpointVerifier {
         // on the next call(). Failures always reject a block, so we know
         // there will be at least one more call().
         //
-        // TODO(teor): retry on failure (low priority, failures should be rare)
-        self.process_checkpoint_range();
+        // We don't retry with a smaller range on failure, because failures
+        // should be rare.
+        let mut valid_blocks = self.process_checkpoint_range();
 
         metrics::gauge!("checkpoint.queued_slots", self.queued.len() as i64);
 
-        async move {
-            // Remove the Result<..., RecvError> wrapper from the channel future
-            rx.await
-                .expect("CheckpointVerifier does not leave dangling receivers")
+        // Optimisation: return a small future in the common case.
+        //
+        // Rust 1.46 and later have known regressions in future size and
+        // complexity, so we want to return a small future in the common case.
+        if valid_blocks.is_empty() {
+            async move {
+                // Remove the Result<..., RecvError> wrapper from the channel future.
+                rx.await
+                    .expect("CheckpointVerifier does not leave dangling receivers")
+            }
+            .boxed()
+        } else {
+            let mut state_service = self.state_service.clone();
+
+            async move {
+                // All the blocks in the current range are valid. Commit them to the
+                // state.
+                for qblock in valid_blocks.drain(..) {
```

Comment:
unless you're working with a `&mut Vec` here I don't think there's any reason to use drain, just iterating over the container will implicitly call `IntoIterator` which in the case of a Vec takes ownership of the values and consumes the container.

```suggestion
                for qblock in valid_blocks {
```

Reply:
I have become too reliant on clippy for hints like this. 🤷